### ISSUE TITLE
Loading screen text customisation by mods

### ISSUE THREAD

**Issue by ClobberXD** [minetest/minetest#9897]
## Problem

There can be games that place schematics when creating a new world. Depending on how large the schematic is, the first load (or coming to think of it, any load really) could be pretty long.

## Solutions

It'd be nice if mods are given access to the loading screen text. Currently, any code that's outside of functions and other scopes are executed at load-time already. To leverage the load screen text customisation functionality, mods could make use of a `register_on_load` callback, like so:

```
minetest.register_on_load("Placing schematics...", function()
    -- Do time-taking stuff here
end)
```

This isn't a global callback per-se - it only looks like one, because there's no "mod registration API" in MT. This callback will run immediately upon being registered, as it's supposed to run at load-time - this sounds very sketchy, but I'm unable to think of a better way.

This callback should run after item and node initialisation (the final stage of engine loading). I assume this is when mods' assets are fully loaded and ready for use, and the environment is ready. I'm hoping this would run *right* before the first server step, so there shouldn't be any issues when working with stuff like `get/set_node`, LVMs, etc. 

This would only be used when a server+client instance is launched (either singleplayer, or a server launched from GUI). Would it be possible for players to join before the server user themselves are connected? Servers technically start accepting connections after "Creating server", right? If so, we need to account for a client attempting to connect before the custom load action is complete. Ideally, the client should get the same custom load string once the cross the "Initialising nodes" stage. I'm pretty sure I'm making too many assumptions for my own good here. :)

## Alternatives

Feel free to suggest alternatives.

## Additional context

Not applicable / can't think of any.

**Comment by SmallJoker** [comment]
Related #9549

**Comment by ClobberXD** [comment]
> Your request makes a backwards jump from 7 to 3, which is somewhat not doable at this point.

Is it because the server environment wouldn't be ready by then? I'm assuming it'd be ready even before stage 1.

**Comment by SmallJoker** [comment]
@ClobberXD ServerEnvironment is ready at stage 1, but the player's position (and hence, positional schematic-placements) will only be available in step 6.

**Comment by ClobberXD** [comment]
The player needn't be added to the environment for this feature - if the player has to be present, the mod would be better off using `on_joinplayer` instead.

The idea is that mods can decide to run a long-running process at server startup - this is already possible, AFAIK, as Lua doesn't run in parallel. But since a long-running process would result in a connecting client apparently "freezing" during load-time, I thought of the possibility of notifying the client via a load screen message.

**Comment by sorcerykid** [comment]
> Currently this can be done by formspecs and attaching the player on join.

Actually that doesn't work because the player can simply hit escape to close the formspec. I attempted that for my [Pluggable Helpers](https://imgur.com/XncbYnT) server, since the entire interface is formspec-based. 

I think the big question is: Why does the player need to be added to the environment in order to display text or graphics on the client's loading screen? The graphics wouldn't even have to be an asset. They could be a dynamically generated texture in base64 encoding a-la #9767, possibly even with a length hard-limit to prevent misuse of the feature.

**Comment by ClobberXD** [comment]
> I think the big question is: Why does the player need to be added to the environment in order to display text or graphics on the client's loading screen?

I really hope that's not the case, and SmallJoker misunderstood my request. :)

The request is all about displaying custom text to the client, when they're waiting for the server to finish setting up a long job. I'm also pretty sure the "waiting for the long job to complete" is already the case, as the server logic is single-threaded, and would have to finish executing all the scripts first, before proceeding further.

This request **does not** ask for the player to be present; rather, the player should only be informed about the ongoing process while waiting mid-connection for the long process to finish running. An example for this would be when a game wants to setup a custom spawn (using schematics or the like) - it'd be nice if players who try to connect while this process is underway are informed about this via custom loading screen text.

I'm really hoping this request doesn't involve flipping around the server initialisation order, and things work out of the box (although that rarely ever happens).

**Comment by SmallJoker** [comment]
> I really hope that's not the case, and SmallJoker misunderstood my request. :)

The problem isn't about the text, but the actions you'd like to perform. What should be executed in the callback mentioned in the first post? It cannot do anything player-specific at this point.

**Comment by ClobberXD** [comment]
I was thinking of more generic long-running actions, like placing schematics, et al. This could be useful for games that want to control exactly how the map should be, like CTF. Now, I understand placing schematics doesn't take *that* long, but it's just an example I could come up with - there could be games that might want to manipulate maps to a much greater extent. Continuing with the example, the schematic that needs to be loaded is decided solely by the mod, in this case, independent of player joins.

TL;DR: All I ask in this request is for environment access to be available, and nothing player-specific.
